sources/ghostdog/src/main.rs

```diff
@@ -0,0 +1,181 @@
+/*!
+ghostdog is a tool to manage ephemeral disks.
+It can be called as a udev helper program to identify ephemeral disks.
+*/
+
+use argh::FromArgs;
+use gptman::GPT;
+use hex_literal::hex;
+use lazy_static::lazy_static;
+use signpost::uuid_to_guid;
+use snafu::ResultExt;
+use std::collections::HashSet;
+use std::fs;
+use std::io::{Read, Seek};
+use std::path::PathBuf;
+
+#[derive(FromArgs, PartialEq, Debug)]
+/// Manage ephemeral disks.
+struct Args {
+    #[argh(subcommand)]
+    subcommand: SubCommand,
+}
+
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand)]
+enum SubCommand {
+    Scan(ScanArgs),
+}
+
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand, name = "scan")]
+/// Scan a device to see if it is an ephemeral disk.
+struct ScanArgs {
+    #[argh(positional)]
+    device: PathBuf,
+}
+
+// Main entry point.
+fn run() -> Result<()> {
+    let args: Args = argh::from_env();
+    match args.subcommand {
+        SubCommand::Scan(scan_args) => {
+            let path = scan_args.device;
+            let mut f = fs::File::open(&path).context(error::DeviceOpen { path })?;
+            let device_type = find_device_type(&mut f)?;
+            emit_device_type(&device_type);
+        }
+    }
+    Ok(())
+}
+
+/// Find the device type by examining the partition table, if present.
+fn find_device_type<R>(reader: &mut R) -> Result<String>
+where
+    R: Read + Seek,
+{
+    // We expect the udev rules to only match block disk devices, so it's fair
+    // to assume it could have a partition table, and that it's probably an
+    // unformatted ephemeral disk if it doesn't.
```

Comment:
Will this tag an unformatted EBS volume as an ephemeral disk?

Reply:
It will, though we won't add a symlink in udev unless the model matches "Amazon EC2 NVMe Instance Storage".

I opted to keep the code here pretty simple. At least some of the filtering has to happen in udev, or else we would execute a `ghostdog` process for all sorts of devices, so I thought it made sense to prefer udev rules where possible.